sklearn/preprocessing/_encoders.py

```diff
@@ -112,7 +112,15 @@ def _transform(self, X, handle_unknown='error'):
                     # continue `The rows are marked `X_mask` and will be
                     # removed later.
                     X_mask[:, i] = valid_mask
-                    Xi = Xi.copy()
+                    # cast Xi into the largest string type
+                    # necessary to handle different lengths
+                    if (isinstance(self.categories_[i][0],
+                                   six.string_types)
+                            and self.categories_[i].itemsize > Xi.itemsize):
```

Comment:
if `self.categories_[i]` has an object dtype, itemsize will always be 8 on a 64 bit system. I don't think that's what you want here either.

Reply:
Indeed, but I think we need to make the first check actually check for string dtype, so if you have object dtype you never get to this itemsize check. 

I am not fully sure what the best way is to check for a numpy string dtype. Maybe `self.categories_.dtype.kind in ('U', 'S')` ?

Reply:
I don't think it actually tests for a string dtype?

```
In [82]: isinstance(np.array(['a', 'b'], dtype='U1')[0], six.string_types)
Out[82]: True

In [83]: isinstance(np.array(['a', 'b'], dtype='object')[0], six.string_types)
Out[83]: True
```

we need a check where only the first gives True.

Reply:
Actually we also check that the sizes of the elements in categories array are larger than the elements in the input matrix. So if the classes are objects the check will not pass.

However I updated the test with a line that checks `type(categories_[0]) != object` I think it is open for discussion

Reply:
@GMarzinotto I don't think `type(categories_[0]) != object` is doing what you want (as even for an object array with string, an element in it is a python string, and thus not equal to `object`)

Why not use ``self.categories_.dtype.kind in ('U', 'S')`` as I mentioned above? (that also saves one line)

Reply:
I have tried this and it fails because sometimes 
categories is a list, and in those cases  

`AttributeError: 'list' object has no attribute 'dtype'`

In can check `instanceof(list)`
However this is getting really complicated for no reason.

as I stated in my previous comment, we also check that the sizes of the elements in categories array are larger than the elements in the input matrix. So if the classes are objects the check will not pass

For me the older version was OK, only checking for six strings and checking the size difference.

Reply:
probably rather isinstance np.ndarray, right?

Reply:
or ``getattr(categories, dtype, None)``?

Reply:
> I have tried this and it fails because sometimes
> categories is a list, and in those cases
>
> AttributeError: 'list' object has no attribute 'dtype'

@GMarzinotto sorry, there was a small type in my suggestion. It is `self.categories_[i].dtype.kind in ('U', 'S')` (note the `[i]`). 
This should always be a ndarray. And you already check the itemsize on it with `self.categories_[i].itemsize`, so that will certainly work.

I think doing that will simplify it, as then you only need 2 checks (the dtype kind + itemsize) instead of the 4 now.